design/proposal-crossplane-v2.md

```diff
@@ -0,0 +1,948 @@
+MRs, like status conditions and management policies. MR Go structs satisfy the
+interface by:
+
+1. Embedding structs from crossplane-runtime (e.g. `xpv1.ResourceSpec`) in their
+   spec and status
+2. Generating getter and setter methods using
+   [crossplane-tools](https://github.com/crossplane/crossplane-tools)
+
+The MR reconciler works with any struct that satisfies the `resource.Managed`
+interface. It orchestrates an `ExternalClient` implementation. An
+`ExternalClient` is a set of create, read, update, and delete (CRUD) methods
+that operate on a specific type of MR struct. Given an MR struct as input, they
+either create, read, update, or delete the resource it represents in an external
+system like AWS.
+
+Upjet providers are the exception to this rule. Upjet MR structs also satisfy
+the `resource.Terraformed` interface, using generated getters and setter
+methods. Upjet providers use a single `ExternalClient` implementation that works
+with any MR Go struct that satisfies this interface.
+
+Given this context, the process to build a namespaced provider would be:
+
+1. Duplicate all MR structs
+    1. Remove the `// +kubebuilder:resource:scope=Cluster` comment marker
+    2. Change the API group constant to end in `m.crossplane.io`
+    3. Change to a namespaced variant of the embedded `xpv1.ResourceSpec`
+    4. Re-run crossplane-tools and Upjet code generation against the updated
+       types
+2. Duplicate all `ExternalClient` implementations
+    5. Change the duplicated implementations to use the duplicated structs
+3. Duplicate main.go and a few “plumbing” files that register types and
+   construct controllers
+
+Upjet providers can skip step 2, as there are no unique `ExternalClient`
+implementations to duplicate.
+
+A namespaced variant of the embedded `xpv1.ResourceSpec` struct is required to
+remove the `spec.writeConnectionSecretRef.namespace` field. Namespaced MRs will
+only write connection secrets to their own namespace.
+
+I propose these changes be made in a branch of existing provider repositories -
+e.g. either by making the changes in a `namespaced` branch, or moving cluster
+scoped development to a `cluster-scope` branch and switching the `main` branch
+to the more modern, namespaced approach.
+
+In Crossplane v1.x the package manager treats `v1.0.0-namespaced` as a
+pre-release of semantic version `v1.0.0`. I propose that Crossplane v2 treat
+hyphenated version suffixes as a “version modifier”.
+
+This is common in official Docker images. For example `golang:1.23.5` and
+`golang:1.23.5-alpine` both install Go 1.23.5, but the former uses a Debian base
+image while the latter uses Alpine.
+
+To do this, the package manager would consider only versions with matching
+suffixes when solving a dependency. For example the constraint `>=
+v1.0.0-namespaced` would be satisfied by `v1.1.0-namespaced`, but not by `v1.1.`
+or `v1.1.0-other`.
+
+#### Introduce `ManagedResourceDefinition`
```

Comment:
I'm cooling off on this idea a bit. Wrote some thoughts at https://github.com/crossplane/crossplane/issues/2262#issuecomment-2723074197.

Is this something anyone is passionate about?